src/pybind/mgr/cephadm/inventory.py

```diff
@@ -66,6 +67,61 @@ def __init__(
         self.hostname = hostname
 
 
+class DaemonDeployQueue:
+    """
+    Queue of daemons cephadm thinks it should deploy
+    Note this only covers fresh deployments, not redeploy/reconfig/restart
+    """
+
+    def __init__(self) -> None:
+        # mapping of hosts to DaemonPlacement objects that are meant
+        # to be deployed on that host
+        self._daemons: Dict[str, List[Tuple[DaemonPlacement, ServiceSpec]]] = {}
+
+    def add_to_queue(self, slots: List[Tuple[DaemonPlacement, ServiceSpec]]) -> None:
+        for dp, spec in slots:
+            if not dp.daemon_type:
+                raise OrchestratorError('Daemon Deploy Queue got DaemonPlacement with no daemon type')
+            if not dp.hostname:
+                raise OrchestratorError('Daemon Deploy Queue got DaemonPlacement with no hostname')
+            if dp.hostname not in self._daemons:
+                self._daemons[dp.hostname] = []
+            self._daemons[dp.hostname].append((dp, spec))
+
+    def remove_from_queue(self, daemon_name: str, hostname: str) -> None:
+        if hostname not in self._daemons:
+            raise OrchestratorError(
+                f'Daemon Deploy Queue got request to remove {daemon_name} from queue host {hostname} with no daemons queued'
+            )
+        self._daemons[hostname] = [(dp, spec) for (dp, spec) in self._daemons[hostname] if dp.name != daemon_name]
+        if not self._daemons[hostname]:
+            self._daemons.pop(hostname, None)
+
+    def get_all_queued_daemons(self) -> List[Tuple[DaemonPlacement, ServiceSpec]]:
```

Comment:
Should the queue type itself be iterable? Add an __iter__method and people could write:
```
for placement, spec in queue:
    ....
```
rather than needing to call this additional method.